### PR TITLE
Adding experimental command to list validators

### PR DESCRIPTION
**Pull request by ilblackdragon**
Ref #188 

Command:
```
near validators current
```

Current output:
```
Validators (total: 54, seat price: 51,205):
.------------------------------------------------------------------------------------.
|     Validator Id      |  Stake  | # seats | % online | bls produced | bls expected |
|-----------------------|---------|---------|----------|--------------|--------------|
| node3                 | 304,940 | 6       |      100 |          360 |          360 |
| node1                 | 304,924 | 6       |      100 |          361 |          361 |
| node0                 | 304,903 | 6       |      100 |          362 |          362 |
| node2                 | 277,171 | 5       |      100 |          361 |          361 |
| huglester.betanet     | 209,874 | 4       |      100 |          288 |          288 |
| hashquark             | 166,381 | 3       |      100 |          217 |          217 |
| unknown.test          | 137,481 | 3       |      100 |          144 |          144 |
| nuc2.test             | 110,512 | 2       |      100 |          145 |          145 |
| fredrikmalmqvist.test | 109,606 | 2       |       99 |          144 |          145 |
| dochpryof.test        | 109,566 | 2       |      100 |          144 |          144 |
| launooskuarttu.test   | 109,159 | 2       |      100 |          145 |          145 |
| janliamnilsson.test   | 108,907 | 2       |       99 |          143 |          144 |
| olaiolsen.test        | 108,806 | 2       |       99 |          144 |          145 |
| bowen.test            | 108,779 | 2       |      100 |          145 |          145 |
| tommywesley.test      | 108,583 | 2       |       98 |          143 |          145 |
| simonhugo.test        | 108,081 | 2       |      100 |          145 |          145 |
| oligarr.test          | 107,980 | 2       |      100 |          144 |          144 |
| felixschulz.test      | 107,478 | 2       |       75 |          109 |          145 |
| staked.test           | 107,033 | 2       |      100 |          144 |          144 |
| bitoven.test          | 106,765 | 2       |      100 |          144 |          144 |
| volvos60909.test      | 105,928 | 2       |      100 |          145 |          145 |
| jaroslavrud.test      | 105,579 | 2       |       59 |           86 |          144 |
| anonstake.test        | 105,525 | 2       |      100 |          144 |          144 |
| isillien.betanet      | 105,093 | 2       |      100 |          145 |          145 |
| pony.test             | 104,468 | 2       |      100 |          145 |          145 |
| buster.betanet        | 104,146 | 2       |      100 |          145 |          145 |
| pathrock.test         | 103,716 | 2       |      100 |          145 |          145 |
| gunray.test           | 102,611 | 2       |      100 |          145 |          145 |
| validatorsonline.test | 102,459 | 2       |      100 |          145 |          145 |
| andreyvelde.test      | 102,415 | 2       |      100 |          144 |          144 |
| unicorn.betanet       | 102,410 | 2       |      100 |          145 |          145 |
| nodeasy.test          | 102,179 | 2       |      100 |           72 |           72 |
| catcatcat.test        | 101,628 | 2       |      100 |           72 |           72 |
| danil.betanet         | 101,005 | 2       |      100 |           73 |           73 |
| zav.betanet           | 100,958 | 2       |      100 |           72 |           72 |
| sll.betanet           | 100,902 | 2       |      100 |           72 |           72 |
| fattox.test           | 100,721 | 2       |      100 |           72 |           72 |
| gaia.test             | 100,721 | 2       |        0 |            0 |           72 |
| stateb.betanet        | 100,717 | 2       |      100 |           72 |           72 |
| sparkpool.test        | 100,596 | 2       |      100 |           72 |           72 |
| moonlet.test          | 100,595 | 2       |      100 |           72 |           72 |
| afi.betanet           | 100,500 | 2       |      100 |           73 |           73 |
| fra.betanet           | 100,500 | 2       |      100 |           72 |           72 |
| max.betanet           | 100,500 | 2       |      100 |           72 |           72 |
| itokenpool.betanet    | 100,402 | 2       |      100 |           73 |           73 |
| harrypotter           | 100,304 | 2       |      100 |           72 |           72 |
| gladba.betanet        | 89,290  | 2       |      100 |           72 |           72 |
| ubikcapital.test      | 85,260  | 2       |      100 |           72 |           72 |
| lxyw.betanet          | 75,730  | 1       |      100 |           72 |           72 |
| zavodil.betanet       | 75,564  | 1       |      100 |           72 |           72 |
| something.betanet     | 75,229  | 1       |      100 |           72 |           72 |
| gamah.betanet         | 75,197  | 1       |        0 |            0 |           72 |
| vitalpointai.betanet  | 70,032  | 1       |      100 |           72 |           72 |
| blaze.betanet         | 55,168  | 1       |      100 |           72 |           72 |
'------------------------------------------------------------------------------------'
```

```
near validators next
```

Output:
```
Next validators (total: 54, seat price: 51,112):
.-------------------------------------------------------------------.
|   Status   |       Validator       |       Stake        | # seats |
|------------|-----------------------|--------------------|---------|
| New        | c1.nearkat            | 80,206             | 2       |
| New        | c29r3.betanet         | 100,499            | 2       |
| New        | kokos.betanet         | 100,000            | 2       |
| Rewarded   | afi.betanet           | 100,500 -> 100,545 | 2       |
| Rewarded   | andreyvelde.test      | 102,415 -> 102,460 | 2       |
| Rewarded   | anonstake.test        | 105,525 -> 105,572 | 2       |
| Rewarded   | bitoven.test          | 106,765 -> 106,812 | 2       |
| Rewarded   | blaze.betanet         | 55,168 -> 55,192   | 1       |
| Rewarded   | bowen.test            | 108,779 -> 108,827 | 2       |
| Rewarded   | buster.betanet        | 104,146 -> 104,192 | 2       |
| Rewarded   | catcatcat.test        | 101,628 -> 101,673 | 2       |
| Rewarded   | danil.betanet         | 101,005 -> 101,050 | 2       |
| Rewarded   | dochpryof.test        | 109,566 -> 109,614 | 2       |
| Rewarded   | fattox.test           | 100,721 -> 100,766 | 2       |
| Rewarded   | felixschulz.test      | 107,478 -> 107,525 | 2       |
| Rewarded   | fra.betanet           | 100,500 -> 100,545 | 2       |
| Rewarded   | fredrikmalmqvist.test | 109,606 -> 109,655 | 2       |
| Rewarded   | gladba.betanet        | 89,290 -> 89,330   | 2       |
| Rewarded   | gunray.test           | 102,611 -> 102,656 | 2       |
| Rewarded   | harrypotter           | 100,304 -> 100,348 | 2       |
| Rewarded   | hashquark             | 166,381 -> 166,455 | 3       |
| Rewarded   | huglester.betanet     | 209,874 -> 209,967 | 4       |
| Rewarded   | isillien.betanet      | 105,093 -> 105,140 | 2       |
| Rewarded   | itokenpool.betanet    | 100,402 -> 100,447 | 2       |
| Rewarded   | janliamnilsson.test   | 108,907 -> 108,955 | 2       |
| Rewarded   | launooskuarttu.test   | 109,159 -> 109,207 | 2       |
| Rewarded   | lxyw.betanet          | 75,730 -> 75,763   | 1       |
| Rewarded   | max.betanet           | 100,500 -> 100,545 | 2       |
| Rewarded   | moonlet.test          | 100,595 -> 100,640 | 2       |
| Rewarded   | node0                 | 304,903 -> 305,037 | 6       |
| Rewarded   | node1                 | 304,924 -> 305,058 | 6       |
| Rewarded   | node2                 | 277,171 -> 277,294 | 5       |
| Rewarded   | node3                 | 304,940 -> 305,074 | 6       |
| Rewarded   | nodeasy.test          | 102,179 -> 102,224 | 2       |
| Rewarded   | nuc2.test             | 110,512 -> 110,561 | 2       |
| Rewarded   | olaiolsen.test        | 108,806 -> 108,854 | 2       |
| Rewarded   | oligarr.test          | 107,980 -> 108,028 | 2       |
| Rewarded   | pathrock.test         | 103,716 -> 103,762 | 2       |
| Rewarded   | pony.test             | 104,468 -> 104,514 | 2       |
| Rewarded   | simonhugo.test        | 108,081 -> 108,129 | 2       |
| Rewarded   | sll.betanet           | 100,902 -> 100,947 | 2       |
| Rewarded   | something.betanet     | 75,229 -> 75,262   | 1       |
| Rewarded   | sparkpool.test        | 100,596 -> 100,640 | 2       |
| Rewarded   | staked.test           | 107,033 -> 107,080 | 2       |
| Rewarded   | stateb.betanet        | 100,717 -> 100,762 | 2       |
| Rewarded   | tommywesley.test      | 108,583 -> 108,631 | 2       |
| Rewarded   | ubikcapital.test      | 85,260 -> 85,297   | 2       |
| Rewarded   | unicorn.betanet       | 102,410 -> 102,456 | 2       |
| Rewarded   | unknown.test          | 137,481 -> 137,542 | 3       |
| Rewarded   | validatorsonline.test | 102,459 -> 102,504 | 2       |
| Rewarded   | vitalpointai.betanet  | 70,032 -> 70,062   | 1       |
| Rewarded   | volvos60909.test      | 105,928 -> 105,974 | 2       |
| Rewarded   | zav.betanet           | 100,958 -> 101,003 | 2       |
| Rewarded   | zavodil.betanet       | 75,564 -> 75,597   | 1       |
| Kicked out | gaia.test             | -                  | -       |
| Kicked out | gamah.betanet         | -                  | -       |
| Kicked out | jaroslavrud.test      | -                  | -       |
'-------------------------------------------------------------------'
```

```
near proposals
```

Output:
```
Proposals for the epoch after next (total: 6, expected seat price = 52,234)
.-----------------------------------------------------------------.
|       Status       |       Validator       | Stake => New Stake |
|--------------------|-----------------------|--------------------|
| Rollover           | node3                 | 304,940            |
| Rollover           | node1                 | 304,924            |
| Rollover           | node0                 | 304,903            |
| Rollover           | node2                 | 277,171            |
| Rollover           | huglester.betanet     | 209,874            |
| Rollover           | hashquark             | 166,381            |
| Proposal(Accepted) | unknown.test          | 137,481 => 165,985 |
| Rollover           | nuc2.test             | 110,512            |
| Rollover           | fredrikmalmqvist.test | 109,606            |
| Rollover           | dochpryof.test        | 109,566            |
| Proposal(Accepted) | ractolechoc5.test     | 109,296            |
| Rollover           | launooskuarttu.test   | 109,159            |
| Proposal(Accepted) | villiamsivertsen.test | 108,964            |
| Rollover           | janliamnilsson.test   | 108,907            |
| Rollover           | olaiolsen.test        | 108,806            |
| Rollover           | bowen.test            | 108,779            |
| Rollover           | tommywesley.test      | 108,583            |
| Rollover           | simonhugo.test        | 108,081            |
| Rollover           | oligarr.test          | 107,980            |
| Rollover           | felixschulz.test      | 107,478            |
| Rollover           | staked.test           | 107,033            |
| Rollover           | bitoven.test          | 106,765            |
| Rollover           | volvos60909.test      | 105,928            |
| Rollover           | jaroslavrud.test      | 105,579            |
| Rollover           | anonstake.test        | 105,525            |
| Rollover           | isillien.betanet      | 105,093            |
| Rollover           | pony.test             | 104,468            |
| Rollover           | buster.betanet        | 104,146            |
| Rollover           | pathrock.test         | 103,716            |
| Rollover           | gunray.test           | 102,611            |
| Rollover           | validatorsonline.test | 102,459            |
| Rollover           | andreyvelde.test      | 102,415            |
| Rollover           | unicorn.betanet       | 102,410            |
| Rollover           | nodeasy.test          | 102,179            |
| Rollover           | catcatcat.test        | 101,628            |
| Rollover           | danil.betanet         | 101,005            |
| Rollover           | zav.betanet           | 100,958            |
| Rollover           | sll.betanet           | 100,902            |
| Rollover           | fattox.test           | 100,721            |
| Rollover           | gaia.test             | 100,721            |
| Rollover           | stateb.betanet        | 100,717            |
| Rollover           | sparkpool.test        | 100,596            |
| Rollover           | moonlet.test          | 100,595            |
| Rollover           | afi.betanet           | 100,500            |
| Rollover           | fra.betanet           | 100,500            |
| Rollover           | max.betanet           | 100,500            |
| Rollover           | itokenpool.betanet    | 100,402            |
| Rollover           | harrypotter           | 100,304            |
| Proposal(Accepted) | stakingfund.test      | 100,000            |
| Rollover           | gladba.betanet        | 89,290             |
| Rollover           | ubikcapital.test      | 85,260             |
| Rollover           | lxyw.betanet          | 75,730             |
| Rollover           | zavodil.betanet       | 75,564             |
| Rollover           | something.betanet     | 75,229             |
| Rollover           | gamah.betanet         | 75,197             |
| Rollover           | vitalpointai.betanet  | 70,032             |
| Rollover           | blaze.betanet         | 55,168             |
| Proposal(Declined) | dokiacapital.betanet  | 11,133             |
| Proposal(Declined) | a.betanet             | 500                |
'-----------------------------------------------------------------'
Expected seat price is calculated based on observed so far proposals and validators.
It can change from new proposals or some validators going offline.
Note: this currently doesn't account for offline kickouts and rewards for current epoch
```

<a href="https://gitpod.io/#https://github.com/nearprotocol/near-shell/pull/328"></a>

### CODE REVIEW

**Comment by mfornet** [comment]
WDYT about sorting the rows by total staked instead of by validator name?

**Comment by ilblackdragon** [comment]
@mfornet done.

File: package.json [hunk 1]
```diff
@@ -31,7 +31,9 @@
     "strip-ansi-cli": "^2.0.0"
   },
   "dependencies": {
+    "ascii-table": "0.0.9",
     "bn": "^1.0.5",
+    "bn.js": "^5.1.1",
```
Comment (ilblackdragon) [inline package.json:36]: I'm actually not sure what bn above is for?... @janedegtiareva you added it here https://github.com/nearprotocol/near-shell/commit/5ad3a891d5d1b6240df78dc87d285f5c6a290c68 but this is not `bn.js`.

**Comment by ilblackdragon** [comment]
Also @bowenwang1996 @frol interestingly, indeed  https://github.com/nearprotocol/nearcore/issues/2259 is not needed, but seat price is actually not everything. We also cut some validators from their seats, because usually with given threshold it will be over 100 seats filled. To reproduce this logic, we actually need to reproduce the random shuffle that our EpochManager does...

File: package.json (hunk 1 above)
Comment (janedegtiareva) [reply package.json:36]: This was most likely refactored out at some point. I don't see any usages, so we can remove the other one. I'll put together a cleanup PR

File: commands/validators.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                Math.floor(validator.num_produced_blocks / validator.num_expected_blocks * 100),
+                validator.num_produced_blocks,
+                validator.num_expected_blocks);
+        });
+        console.log(validatorsTable.toString());
+
+        if (validators.current_fishermen) {
+            console.log(`\nFishermen (total: ${validators.current_fishermen.length}):`);
+            var fishermenTable = new AsciiTable();
+            fishermenTable.setHeading('Fisherman Id', 'Stake');
```
Comment (janedegtiareva) [inline commands/validators.js:94]: What do you think of adding a little text re what is a fisherman? or at least a link to docs.

File: commands/validators.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
+const exitOnError = require('../utils/exit-on-error');
+const connect = require('../utils/connect');
+const inspectResponse = require('../utils/inspect-response');
+const bs58 = require('bs58');
+const { utils } = require('near-api-js');
+const BN = require('bn.js');
+const AsciiTable = require('ascii-table');
+
+function findSeatPrice(validators, numSeats) {
```
Comment (vgrichina) [inline commands/validators.js:7]: this has way too many logic without explanation of what it is trying to do and why

also sounds like stuff like this should live in near-api-js, near-shell only taking care of displaying it as a table, etc

File: commands/validators.js (hunk 2 above)
Comment (ilblackdragon) [reply commands/validators.js:90]: Seems like a weird place IMO. We don't have explanation for anything else, do we?
Like what is transaction, staking, validators is as unclear as fishermen.

File: commands/validators.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+function diffValidators(currentValidators, nextValidators) {
+    const result = { newValidators: [], removedValidators: [], changeValidators: [] };
+    const validatorsMap = new Map();
+    const nextValidatorsMap = new Map();
+    nextValidators.map((validator) => nextValidatorsMap[validator.account_id] = validator);
+    currentValidators.map((validator) => { 
+        if (!(validator.account_id in nextValidatorsMap)) {
+            result.removedValidators.push(validator);
```
Comment (vgrichina) [inline commands/validators.js:39]: why use `.map` and then `.push` results somewhere? This is very confusing, makes sense to use `for (const validator of currentValidators)` or `.forEach` at least.

File: commands/validators.js (hunk 3 above)
Comment (ilblackdragon) [reply commands/validators.js:7]: see https://github.com/nearprotocol/near-shell/pull/328#issuecomment-619889936

File: commands/validators.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    });
+    nextValidators.map((validator) => {
+        if (validator.account_id in validatorsMap) {
+            if (validatorsMap[validator.account_id].stake != validator.stake) {
+                result.changeValidators.push({new: validator, old: validatorsMap[validator.account_id]});
+            }
+        } else {
+            result.newValidators.push(validator);
+        }
+    });
```
Comment (vgrichina) [inline commands/validators.js:51]: seems like whole thing could be refactored like this:

```js
const nextValidatorsSet = new Set(nextValidators.map(v => v.account_id));
const removedValidators = currentValidators.filter(v => nextValidatorsSet.has(v.account_id));
const newValidators = nextValidators.filter(v => !validatorsSet.has(v.account_id));
// ... etc
```

in current code it's kinda hard to follow data flow and what is filtered where. E.g. is it intentional that in `nextValidators.map((validator) => {` loop some validators go neither to `changeValidators` nor to `newValidators`?

File: commands/validators.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    return result;
+}
+
+module.exports = {
+    command: 'EXPERIMENTAL_validators',
```
Comment (vgrichina) [inline commands/validators.js:56]: seems like a bit suboptimal to prefix commands with allcaps `EXPERIMENTAL_` thing. How about we just use `_` as prefix (similar to "private" variables / methods in JS).

File: commands/validators.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    desc: 'lookup validators and proposals',
+    handler: exitOnError(async (argv) => {
```
Comment (vgrichina) [inline commands/validators.js:58]: same as with `findSeatPrice` looks like too much logic lives in `near-shell` instead of `near-api-js`. Like if I was building web app to do the same thing, I'll probably need the same logic minus the `AsciiTable`.

File: commands/validators.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+                fishermenTable.addRow(fisherman.account_id, utils.format.formatNearAmount(fisherman.stake, 0));
+            });
+            console.log(fishermenTable.toString());
+        }
+
+        const diff = diffValidators(validators.current_validators, validators.next_validators);
+        console.log(`\nNext validators (total: ${validators.next_validators.length}, seat price: ${utils.format.formatNearAmount(nextSeatPrice, 0)}):`);
+        let nextValidatorsTable = new AsciiTable();
+        nextValidatorsTable.setHeading('Status', 'Validator', 'Stake', '# seats');
+        diff.newValidators.map((validator) => nextValidatorsTable.addRow(
```
Comment (vgrichina) [inline commands/validators.js:101]: let's not use `.map` as `.forEach`

File: commands/validators.js (hunk 3 above)
Comment (vgrichina) [reply commands/validators.js:7]: @ilblackdragon add some comment + TODO with link to GH issue then? Easy to lose context otherwise.

**Comment by jimmy3dita** [comment]
Super welcome feature! I feel limited by the structure of the CLI, mostly because it can only see "the now" and not in the past.

Example of needed features:
- As a CLI user, I want to know the per-seat cost, such that I don't risk to stake a too small amount
- As a CLI user, I want to know if my stake transaction was included in the proposals, such that I know that I will become validator
- As a CLI user, I want to know if my "state" is [no stake, proposal, validator, future validator, future kicked out, past kicked out], such that I can plan my actions accordingly
- As a CLI user, I want to know my "rewards" and performance from the last epoch, such that I can easily report my rewards

On the other hand, I'm not sure that displaying the whole Validator RPC endpoint would be helpful without some time-related queries (e.g., look at past blocks, to know my node performance over time). From this point of view, I'd rather build an application to capture the time series, trying to catch the anomalies.

File: commands/validators.js (hunk 6 above)
Comment (ilblackdragon) [reply commands/validators.js:56]: hm, was following our RPC standard for experimental APIs by @frol .
What is the common way for experimental CLI flags?

Comment (frol) [reply commands/validators.js:56]: It is common to have `experimental` prefix for CLI arguments (well, it is usually in lowercase):

* Node.js: https://nodejs.org/api/cli.html#cli_experimental_import_meta_resolve
* Kubectl: https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/

Docker has special config and env variable to enable experimental CLI arguments: https://github.com/docker/docker.github.io/issues/5718

P.S. it is standard to have kebab-case naming for the CLI arguments (for some odd reason we use everything but kebab-case in near-shell)

Comment (vgrichina) [reply commands/validators.js:56]: > P.S. it is standard to have kebab-case naming for the CLI arguments (for some odd reason we use everything but kebab-case in near-shell)
 
@frol it's not immediately clear, but near-shell does support kebab-case arguments (translated automatically by `yargs`). 

as for having `experimental` prefix – my only concern is that it's not fun to type (especially in caps), not feeling super strongly. 

Also given outstanding proposal to redesign shell CLI, maybe it makes sense to use `near experimental validators`?  
Also does it really have to be "experimental"? Does it mean we can remove it or smth else?

Comment (frol) [reply commands/validators.js:56]: > it's not immediately clear, but near-shell does support kebab-case arguments 

:thinking: 

```
$ near --help
...
  --nodeUrl, --node_url      NEAR node URL  [string] [default: "http://localhost:3030"]
  --networkId, --network_id  NEAR network ID, allows using different keys based on network  [string] [default: "default"]
  --helperUrl                NEAR contract helper URL  [string]
  --keyPath                  Path to master account key  [string]
  --accountId, --account_id  Unique identifier for the account  [string]
```

Do you mean that even though the help message suggests camelCase and snake_case, it will be happy with kebab-case there as well?

> as for having experimental prefix – my only concern is that it's not fun to type (especially in caps), not feeling super strongly.

Well, we definitely don't want to keep it like that for ages, but it is a great way to allow users to use the latest API while allowing us to iterate quickly whenever we feel that changes are needed.

We cannot stabilize CLI feature before we stabilize the RPC endpoint. This implementation uses `EXPERIMENTAL_genesis_config` RPC, which is not settled due to the various changes to the genesis config structure (the most recent is staking storage, I believe).

> Also given outstanding proposal to redesign shell CLI, maybe it makes sense to use near experimental validators?

Looks good to me.

> Also does it really have to be "experimental"?

Well, given `EXPERIMENTAL_genesis_config` may introduce a breaking change, I consider this experimental. API stabilization is going to be a separate effort.

> Does it mean we can remove it or smth else?

Unfortunately, it does.

Comment (vgrichina) [reply commands/validators.js:56]: > Do you mean that even though the help message suggests camelCase and snake_case, it will be happy with kebab-case there as well?

yes

> Well, given EXPERIMENTAL_genesis_config may introduce a breaking change, I consider this experimental. API stabilization is going to be a separate effort.

got it

**Comment by bowenwang1996** [comment]
> To reproduce this logic, we actually need to reproduce the random shuffle that our EpochManager does... In this PR seat number is not always correct and doesn't add up to 100.

@ilblackdragon do we actually want to reproduce this logic? It seems very difficult. We not only need to implement rust's random number generator in javascript, but also find the random seed used there, which means we also need to reproduce the logic of epoch id computation. I suggest that we not do it.

File: commands/validators.js (hunk 5 above)
Comment (ilblackdragon) [reply commands/validators.js:51]: haha, the etc is the hard part :)

**Comment by ilblackdragon** [comment]
@vgrichina Moving into https://github.com/near/near-api-js/pull/300 

@bowenwang1996 Well, we need to make it reproducible for other clients to be able to reproduce... If we don't think our current random function is reproducible (and I looked into it - it's very complicated) - we should change it.

File: commands/validators.js (hunk 5 above)
Comment (vgrichina) [reply commands/validators.js:51]: @ilblackdragon given that this change looks super useful – maybe makes sense to do some minimum cleanup first and release new near-shell:
1) add comment explaining what is happening / link to relevant near-api-js work
2) don't use `.map` as `.forEach` (i.e. just replace with `.forEach` or `for (const item of collection)` loops)
3) don't use CAPS in command

**Comment by bowenwang1996** [comment]
> If we don't think our current random function is reproducible (and I looked into it - it's very complicated) - we should change it.

Agree. Probably some simple shuffling would suffice.

File: commands/validators.js (hunk 6 above)
Comment (ilblackdragon) [reply commands/validators.js:56]: @vgrichina in my case experimental meant that there is no guarantee we won't change command. As you mentioned - we are redesigning command line interface, and I'm not 100% sure how this command should look ilke.

File: package.json [hunk 9]
```diff
@@ -31,14 +31,16 @@
...
     "bs58": "^4.0.1",
     "chalk": "^4.0.0",
     "flagged-respawn": "^1.0.1",
     "is-ci": "^2.0.0",
     "jest-environment-node": "^26.0.0",
     "mixpanel": "^0.11.0",
     "ncp": "^2.0.0",
-    "near-api-js": "^0.23.1",
+    "near-api-js": "git+https://github.com/near/near-api-js.git",
```
Comment (vgrichina) [inline package.json:43]: @ilblackdragon needs fix. When you test locally use https://classic.yarnpkg.com/en/docs/cli/link/

Comment (ilblackdragon) [reply package.json:43]: Yep, waiting for the new release.
